File: test/transpiler/pheno_pkg/src/cs/shootnumber.cs
using System;
using System.Collections.Generic;
public class Shootnumber_
{
    public static Tuple<double,double,List<int>,List<double>,int>  shootnumber_(double canopyShootNumber,double leafNumber,int sowingDensity,double targetFertileShoot,List<double> tilleringProfile,List<int> leafTillerNumberArray,int tillerNumber)
    {
        //- Description:
    //            - Model Name: CalculateShootNumber Model
    //            - Author: Pierre MARTRE
    //            - Reference: Modeling development phase in the 
    //                Wheat Simulation Model SiriusQuality.
    //                See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
    //            - Institution: INRA/LEPSE Montpellier
    //            - Abstract: calculate the shoot number and update the related variables if needed
        //- inputs:
    //            - name: canopyShootNumber
    //                          - min : 0
    //                          - default : 288.0
    //                          - max : 10000
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : shoot m-2
    //                          - description : shoot number for the whole canopy
    //            - name: leafNumber
    //                          - min : 0
    //                          - default : 0
    //                          - max : 10000
    //                          - variablecategory : state
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : leaf
    //                          - description : Leaf number 
    //            - name: sowingDensity
    //                          - parametercategory : constant
    //                          - min : 0
    //                          - datatype : INT
    //                          - max : 500
    //                          - default : 288
    //                          - inputtype : parameter
    //                          - unit : plant m-2
    //                          - description : number of plant /m²
    //            - name: targetFertileShoot
    //                          - min : 280
    //                          - default : 600
    //                          - max : 1000
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLE
    //                          - inputtype : variable
    //                          - unit : shoot
    //                          - description : max value of shoot number for the canopy
    //            - name: tilleringProfile
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLELIST
    //                          - default : [288.0]
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description :  store the amount of new tiller created at each time a new tiller appears
    //            - name: leafTillerNumberArray
    //                          - variablecategory : auxiliary
    //                          - datatype : INTLIST
    //                          - default : [1]
    //                          - inputtype : variable
    //                          - unit : leaf
    //                          - description : store the number of tiller for each leaf layer
    //            - name: tillerNumber
    //                          - min : 0
    //                          - default : 1
    //                          - max : 10000
    //                          - variablecategory : auxiliary
    //                          - datatype : INT
    //                          - inputtype : variable
    //                          - unit : 
    //                          - description :  Number of tiller which appears
        //- outputs:
    //            - name: averageShootNumberPerPlant
    //                          - datatype : DOUBLE
    //                          - min : 0
    //                          - variablecategory : auxiliary
    //                          - max : 10000
    //                          - unit : shoot m-2
    //                          - description : average shoot number per plant in the canopy
    //            - name: canopyShootNumber
    //                          - datatype : DOUBLE
    //                          - min : 0
    //                          - variablecategory : auxiliary
    //                          - max : 10000
    //                          - unit : shoot m-2
    //                          - description : shoot number for the whole canopy
    //            - name: leafTillerNumberArray
    //                          - variablecategory : auxiliary
    //                          - datatype : INTLIST
    //                          - unit : leaf
    //                          - description : store the number of tiller for each leaf layer
    //            - name: tilleringProfile
    //                          - variablecategory : auxiliary
    //                          - datatype : DOUBLELIST
    //                          - unit : 
    //                          - description :  store the amount of new tiller created at each time a new tiller appears
    //            - name: tillerNumber
    //                          - datatype : INT
    //                          - min : 0
    //                          - variablecategory : auxiliary
    //                          - max : 10000
    //                          - unit : 
    //                          - description :  store the amount of new tiller created at each time a new tiller appears
        double averageShootNumberPerPlant;
        double oldCanopyShootNumber;
        int emergedLeaves;
        int shoots;
        int i;
        oldCanopyShootNumber = canopyShootNumber;
        emergedLeaves = (int)(Math.Max(1.0d, Math.Ceiling(leafNumber - 1)));
        shoots = fibonacci_(emergedLeaves).result;
        canopyShootNumber = Math.Min((double)(shoots * sowingDensity), targetFertileShoot);
        averageShootNumberPerPlant = canopyShootNumber / sowingDensity;
        if (canopyShootNumber != oldCanopyShootNumber)
        {
            tilleringProfile.Add(canopyShootNumber - oldCanopyShootNumber);
        }
        tillerNumber = tilleringProfile.Count();
        for (i=leafTillerNumberArray.Count() ; i<(int)(Math.Ceiling(leafNumber)) ; i+=1)
        {
            leafTillerNumberArray.Add(tillerNumber);
        }
        return Tuple.Create(averageShootNumberPerPlant, canopyShootNumber, leafTillerNumberArray, tilleringProfile, tillerNumber);
    }
}

File: src/pycropml/transpiler/antlr_py/csharp/examples/pheno_pkg/src/py/ShootNumber.py
# coding: utf8
from pycropml.units import u
from copy import copy
from array import array

import numpy
from math import *

def model_ShootNumber(canopyShootNumber_t1 = 288.0,
         leafNumber = 3.34,
         sowingDensity = 288.0,
         targetFertileShoot = 600.0,
         tilleringProfile_t1 = [288.0],
         leafTillerNumberArray_t1 = [1, 1, 1],
         numberTillerCohort_t1 = 1):
    """
     - Name: ShootNumber -Version: 1.0, -Time step: 1
     - Description:
                 * Title: CalculateShootNumber Model
                 * Author: Pierre MARTRE
                 * Reference: Modeling development phase in the 
                     Wheat Simulation Model SiriusQuality.
                     See documentation at http://www1.clermont.inra.fr/siriusquality/?page_id=427
                 * Institution: INRA/LEPSE Montpellier
                 * Abstract: calculate the shoot number and update the related variables if needed
     - inputs:
                 * name: canopyShootNumber_t1
                               ** description : shoot number for the whole canopy
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 10000
                               ** default : 288.0
                               ** unit : shoot m-2
                               ** inputtype : variable
                 * name: leafNumber
                               ** description : Leaf number 
                               ** variablecategory : state
                               ** inputtype : variable
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 10000
                               ** default : 3.34
                               ** unit : leaf
                 * name: sowingDensity
                               ** description : number of plant /m²
                               ** parametercategory : species
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 500
                               ** default : 288.0
                               ** unit : plant m-2
                               ** inputtype : parameter
                 * name: targetFertileShoot
                               ** description : max value of shoot number for the canopy
                               ** parametercategory : species
                               ** datatype : DOUBLE
                               ** min : 280
                               ** max : 1000
                               ** default : 600.0
                               ** unit : shoot
                               ** inputtype : variable
                 * name: tilleringProfile_t1
                               ** description :  store the amount of new tiller created at each time a new tiller appears
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** default : [288.0]
                               ** unit : 
                               ** inputtype : variable
                 * name: leafTillerNumberArray_t1
                               ** description : store the number of tiller for each leaf layer
                               ** variablecategory : state
                               ** datatype : INTLIST
                               ** unit : leaf
                               ** default : [1, 1, 1]
                               ** inputtype : variable
                 * name: numberTillerCohort_t1
                               ** description :  Number of tiller which appears
                               ** variablecategory : state
                               ** datatype : INT
                               ** min : 0
                               ** max : 10000
                               ** default : 1
                               ** unit : 
                               ** inputtype : variable
     - outputs:
                 * name: averageShootNumberPerPlant
                               ** description : average shoot number per plant in the canopy
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 10000
                               ** unit : shoot m-2
                 * name: canopyShootNumber
                               ** description : shoot number for the whole canopy
                               ** variablecategory : state
                               ** datatype : DOUBLE
                               ** min : 0
                               ** max : 10000
                               ** unit : shoot m-2
                 * name: leafTillerNumberArray
                               ** description : store the number of tiller for each leaf layer
                               ** variablecategory : state
                               ** datatype : INTLIST
                               ** unit : leaf
                 * name: tilleringProfile
                               ** description :  store the amount of new tiller created at each time a new tiller appears
                               ** variablecategory : state
                               ** datatype : DOUBLELIST
                               ** unit : dimensionless
                 * name: numberTillerCohort
                               ** description : Number of tiller which appears
                               ** variablecategory : state
                               ** datatype : INT
                               ** min : 0
                               ** max : 10000
                               ** unit : dimensionless
    """

    leafTillerNumberArray = []
    tilleringProfile = []
    lNumberArray_rate = []
    emergedLeaves = max(1, ceil(leafNumber - 1.0))
    shoots = fibonacci(emergedLeaves)
    canopyShootNumber = min(shoots * sowingDensity, targetFertileShoot)
    averageShootNumberPerPlant = canopyShootNumber / sowingDensity
    tilleringProfile = copy(tilleringProfile_t1)
    if canopyShootNumber != canopyShootNumber_t1:
        tilleringProfile.append(canopyShootNumber - canopyShootNumber_t1)
    numberTillerCohort = len(tilleringProfile)
    for i in range(len(leafTillerNumberArray_t1) , ceil(leafNumber) , 1):
        lNumberArray_rate.append(numberTillerCohort)
    leafTillerNumberArray = copy(leafTillerNumberArray_t1)
    leafTillerNumberArray.extend(lNumberArray_rate)
    return (averageShootNumberPerPlant, canopyShootNumber, leafTillerNumberArray, tilleringProfile, numberTillerCohort)

def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n - 1) + fibonacci(n - 2)

def init_ShootNumber(sowingDensity = 288.0,
         targetFertileShoot = 600.0):
    tilleringProfile_t1 = []
    leafTillerNumberArray_t1 = []
    leafTillerNumberArray = []
    tilleringProfile = []
    canopyShootNumber = sowingDensity
    averageShootNumberPerPlant = 1.0
    tilleringProfile.append(sowingDensity)
    numberTillerCohort = 1
    return (averageShootNumberPerPlant, canopyShootNumber, leafTillerNumberArray, tilleringProfile, numberTillerCohort)